### PR TITLE
Check for updates

### PR DESCRIPTION
**Pull request by madanadit**
This PR introduces the ability to detect the latest version of Alluxio available and print a message in `master.out` if the version does not match the deployed version.

### CODE REVIEW

**Comment by madanadit** [comment]
jenkins, test this please.

File: core/common/src/main/java/alluxio/util/EnvironmentUtils.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+public final class EnvironmentUtils {
+  private static final Logger LOG = LoggerFactory.getLogger(EnvironmentUtils.class);
+
+  /**
+   * Utility to detect the docker deployment environment.
+   *
+   * @return true, if running on docker
+   */
+  @SuppressFBWarnings("DMI_HARDCODED_ABSOLUTE_FILENAME")
+  public static boolean isDocker() throws IOException {
```
Comment (gpang) [inline core/common/src/main/java/alluxio/util/EnvironmentUtils.java]: do we want these status methods to throw an exception?

File: core/transport/src/proto/journal/journal.proto [hunk 2]
```diff
@@ -24,6 +24,7 @@ message JournalEntry {
   optional AsyncPersistRequestEntry async_persist_request = 16;
   optional BlockContainerIdGeneratorEntry block_container_id_generator = 3;
   optional BlockInfoEntry block_info = 4;
+  optional string cluster_id = 42;
```
Comment (gpang) [inline core/transport/src/proto/journal/journal.proto]: should this be something like a "ClusterInfo" type, which contains the string `cluster_id` and potentially other fields in the future, like start time or something?

File: core/server/master/src/main/java/alluxio/master/meta/UpdateCheck.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    // Create the GET request.
+    Joiner joiner = Joiner.on("/");
+    String path = joiner.join("v0", "version");
+    String url = new URL(new URL(ProjectConstants.UPDATE_CHECK_HOST), path).toString();
+
+    HttpGet post = new HttpGet(url);
+    post.setHeader("User-Agent", getEnvString(clusterID));
+    post.setHeader("Authorization", "Basic " + ProjectConstants.UPDATE_CHECK_AUTH_STRING);
+
+    // Fire off the upload request.
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/meta/UpdateCheck.java]: NIT: not sure what the upload means, so maybe "fire off the version check request"

File: core/server/master/src/main/java/alluxio/master/meta/UpdateCheck.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  /**
+   * @param clusterID the cluster ID
+   * @return a string representation of the user's environment in the format "key1:value1, key2:
+   *         value2".
+   */
+  private static String getEnvString(String clusterID) throws IOException {
+    Joiner joiner = Joiner.on(",");
+    return joiner.join(
+        "ClusterID:" + clusterID,
+      "AlluxioVersion:" + ProjectConstants.VERSION,
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/meta/UpdateCheck.java]: is this indent wrong?

File: core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java [hunk 5]
```diff
@@ -147,6 +149,61 @@ public Address getFieldValue(MasterInfo o) {
...
+     * @param ctx the journal context
+     * @param entry the clusterID journal entry
+     */
+    public void applyAndJournal(java.util.function.Supplier<JournalContext> ctx, String entry) {
+      applyAndJournal(ctx, Journal.JournalEntry.newBuilder().setClusterId(entry).build());
+    }
+
+    @Override
+    public void resetState() {
+      mClusterID = null;
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java]: Should we have a constant id which represents the reset one (instead of using `null`)?

File: core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java [hunk 6]
```diff
@@ -238,6 +296,26 @@ public void start(Boolean isPrimary) throws IOException {
             ThreadFactoryUtils.build("DailyMetadataBackup-%d", true)), mUfs);
         mDailyBackup.start();
       }
+      if (ServerConfiguration.getBoolean(PropertyKey.MASTER_UPDATE_CHECK_ENABLED)) {
+        if (mState.getClusterID() == null) {
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java]: I feel like a cluster id is generally useful, so could this be independent from the version check?

File: core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java [hunk 7]
```diff
@@ -238,6 +296,26 @@ public void start(Boolean isPrimary) throws IOException {
...
+          try (JournalContext context = createJournalContext()) {
+            String clusterID = java.util.UUID.randomUUID().toString();
+            mState.applyAndJournal(context, clusterID);
+            LOG.info("Created new cluster ID {}", clusterID);
+            try {
+              String latestVersion = UpdateCheck.getLatestVersion(clusterID);
+              if (!latestVersion.equals(ProjectConstants.VERSION)) {
+                System.out.println("The latest version (" + latestVersion + ") is not the same"
+                    + "as the current version. To upgrade visit https://www.alluxio.io/download/.");
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java]: print out the current version too

File: core/server/master/src/main/java/alluxio/master/meta/UpdateCheck.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    String path = joiner.join("v0", "version");
+    String url = new URL(new URL(ProjectConstants.UPDATE_CHECK_HOST), path).toString();
+
+    HttpGet post = new HttpGet(url);
+    post.setHeader("User-Agent", getEnvString(clusterID));
+    post.setHeader("Authorization", "Basic " + ProjectConstants.UPDATE_CHECK_AUTH_STRING);
+
+    // Fire off the upload request.
+    HttpClient client = HttpClientBuilder.create().build();
+    HttpResponse response = client.execute(post);
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/meta/UpdateCheck.java]: what happens if the endpoint is unreachable? Shouldn't there be some timeout, so servers don't wait too long?

File: core/common/src/main/java/alluxio/util/EnvironmentUtils.java (hunk 1 above)
Comment (madanadit) [reply core/common/src/main/java/alluxio/util/EnvironmentUtils.java]: we catch all exceptions in the metamaster update check code. i feel that's better than hiding the exception and choosing a default

File: core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java (hunk 5 above)
Comment (madanadit) [reply core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java]: sgtm

File: core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java (hunk 6 above)
Comment (madanadit) [reply core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java]: sgtm

File: core/server/master/src/main/java/alluxio/master/meta/UpdateCheck.java (hunk 4 above)
Comment (madanadit) [reply core/server/master/src/main/java/alluxio/master/meta/UpdateCheck.java]: yes

File: core/transport/src/proto/journal/journal.proto (hunk 2 above)
Comment (madanadit) [reply core/transport/src/proto/journal/journal.proto]: sgtm

File: core/server/master/src/main/java/alluxio/master/meta/UpdateCheck.java (hunk 8 above)
Comment (madanadit) [reply core/server/master/src/main/java/alluxio/master/meta/UpdateCheck.java]: let me explicitly define the retry / timeout params. it does fail when the endpoint is unreachable.

File: core/server/master/src/main/java/alluxio/master/meta/UpdateCheck.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+    // Check the response code.
+    int responseCode = response.getStatusLine().getStatusCode();
+    if (responseCode != HttpURLConnection.HTTP_OK) {
+      throw new IOException("Update check request failed with code: " + responseCode);
```
Comment (haoyuan) [inline core/server/master/src/main/java/alluxio/master/meta/UpdateCheck.java]: Does this mean that if the HTTP request failed, the user would see exception? If yes, suggest to make it silent. 

@apc999 @madanadit

Comment (madanadit) [reply core/server/master/src/main/java/alluxio/master/meta/UpdateCheck.java]: There will be a debug log level message in `master.log`

**Comment by madanadit** [comment]
@gpang PTAL

**Comment by madanadit** [comment]
jenkins, test this please.

File: core/common/src/main/java/alluxio/util/EnvironmentUtils.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    return new File("/.dockerenv").exists();
```
Comment (apc999) [inline core/common/src/main/java/alluxio/util/EnvironmentUtils.java]: `alluxio.util.io.FileUtils#exists` which does not throw IOException ?

File: core/server/master/src/main/java/alluxio/master/meta/UpdateCheck.java [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+
+    return EntityUtils.toString(response.getEntity(), "UTF-8");
+  }
+
+  /**
+   * @param clusterID the cluster ID
+   * @return a string representation of the user's environment in the format "key1:value1, key2:
+   *         value2".
+   */
+  private static String getEnvString(String clusterID) throws IOException {
```
Comment (apc999) [inline core/server/master/src/main/java/alluxio/master/meta/UpdateCheck.java]: `getUserAgentString`?

File: core/server/master/src/main/java/alluxio/master/meta/UpdateCheck.java [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+    Joiner joiner = Joiner.on(",");
+    return joiner.join(
```
Comment (apc999) [inline core/server/master/src/main/java/alluxio/master/meta/UpdateCheck.java]: how about following format here:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/User-Agent

File: core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java [hunk 13]
```diff
@@ -229,15 +296,41 @@ public void start(Boolean isPrimary) throws IOException {
...
+      if (mState.getClusterID() == INVALID_CLUSTER_ID) {
+        try (JournalContext context = createJournalContext()) {
+          String clusterID = java.util.UUID.randomUUID().toString();
+          mState.applyAndJournal(context, clusterID);
+          LOG.info("Created new cluster ID {}", clusterID);
+        }
+        if (ServerConfiguration.getBoolean(PropertyKey.MASTER_UPDATE_CHECK_ENABLED)) {
+          try {
+            String latestVersion = UpdateCheck.getLatestVersion(mState.getClusterID(),
+                FormatUtils.parseTimeSize("60sec"),
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java]: why are we using `FormatUtils.parseTimeSize`. Can't we just have a constant long for the time? Also, all these timeouts are a minute. Are they too long? Would this mean startup could take up to 3 additional minutes in the worst case?

File: core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java [hunk 14]
```diff
@@ -229,15 +296,41 @@ public void start(Boolean isPrimary) throws IOException {
...
+                FormatUtils.parseTimeSize("60sec"),
+                FormatUtils.parseTimeSize("60sec")
+            );
+            if (!latestVersion.equals(ProjectConstants.VERSION)) {
+              System.out.println("The latest version (" + latestVersion + ") is not the same"
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java]: this will need a space at the end of the string

File: core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java (hunk 13 above)
Comment (madanadit) [reply core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java]: reduced each of the timeouts to 3 secs

**Comment by madanadit** [comment]
@gpang @apc999 PTAL. Thanks for the comments

**Comment by madanadit** [comment]
jenkins, test this please.

**Comment by madanadit** [comment]
jenkins, test this please.

**Comment by madanadit** [comment]
@apc999 @gpang PTAL

**Comment by madanadit** [comment]
jenkins, test this please.

File: core/server/common/src/main/java/alluxio/master/journal/JournalEntryAssociation.java [hunk 15]
```diff
@@ -52,7 +52,8 @@ public static String getMasterForEntry(JournalEntry entry) {
         || entry.hasBlockInfo()) {
       return Constants.BLOCK_MASTER_NAME;
     }
-    if (entry.hasPathProperties()
+    if (entry.hasClusterInfo()
```
Comment (apc999) [inline core/server/common/src/main/java/alluxio/master/journal/JournalEntryAssociation.java]: it seems 2.0.1 journal will break journal backwards compatibility by throwing `IllegalStateException`

File: core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java [hunk 16]
```diff
@@ -492,18 +582,25 @@ public CheckpointName getCheckpointName() {
...
   public Iterator<Journal.JournalEntry> getJournalEntryIterator() {
-    return mPathProperties.getJournalEntryIterator();
+    return com.google.common.collect.Iterators.concat(mPathProperties.getJournalEntryIterator(),
+        mState.getJournalEntryIterator());
   }
 
   @Override
   public boolean processJournalEntry(Journal.JournalEntry entry) {
-    return mPathProperties.processJournalEntry(entry);
+    return mState.processJournalEntry(entry) || mPathProperties.processJournalEntry(entry);
```
Comment (apc999) [inline core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java]: is this change related to this PR?

Comment (madanadit) [reply core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java]: yes, its to process `clusterInfo` in `mState`

**Comment by madanadit** [comment]
Note: Do NOT cherry-pick to branch-2.0. A companion PR () directed to branch

**Comment by madanadit** [comment]
alluxio-bot, merge this please.

File: core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java [hunk 17]
```diff
@@ -229,15 +295,39 @@ public void start(Boolean isPrimary) throws IOException {
...
+          String clusterID = java.util.UUID.randomUUID().toString();
+          mState.applyAndJournal(context, clusterID);
+          LOG.info("Created new cluster ID {}", clusterID);
+        }
+        if (Boolean.valueOf(ProjectConstants.UPDATE_CHECK_ENABLED)
+            && ServerConfiguration.getBoolean(PropertyKey.MASTER_UPDATE_CHECK_ENABLED)) {
+          try {
+            String latestVersion =
+                UpdateCheck.getLatestVersion(mState.getClusterID(), 3000, 3000, 3000);
+            if (!latestVersion.equals(ProjectConstants.VERSION)) {
```
Comment (apc999) [inline core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java]: `ProjectConstants.VERSION.equals(latestVersion)` in case `lastestVersion` is null?

**Comment by madanadit** [comment]
@apc999 PTAL

**Comment by apc999** [comment]
one minor comment left. Otherwise, LGTM

**Comment by apc999** [comment]
alluxio-bot, merge this please
